python/ray/dag/compiled_dag_node.py

```diff
@@ -355,6 +456,14 @@ def __init__(
         # and the result of a COMPUTE operation will be used by a WRITE operation.
         self._intermediate_buffer: Any = None
 
+        # Store the intermediate result for streamed execution.
+        # The key is the operation, and the value is the input to this operation,
+        # i.e., the result from the last operation. There are only two cases:
+        # if the operation is COMPUTE, the value is the result from the READ with
+        # the same exec_task_idx; if the operation is WRITE, the value is the result
+        # from the COMPUTE with the same exec_task_idx.
+        self._stream_buffer: Dict[_DAGNodeOperation, Any] = {}
```

Comment:
Can we reuse `self._intermediate_buffer` instead?

Reply:
I originally reused it, but for Dict it has additional overhead of hashing/lookup etc. (although arguably the overhead is small), so kept the original path entirely intact.

This also ties to the decision whether we reuse `do_exec_task` or have the new `do_stream_task`.

Reply:
Hmm my preference is that we reuse `do_exec_task` and try to do the `Future` kind of API that @rkooo567 suggested. I think with these changes it should be possible to introduce the stream execution in a clean way.